### ISSUE TITLE
Throughput of kafka-rest

### ISSUE THREAD

**Issue by xiangfu0** [confluentinc/kafka-rest#93]
Hi team, 

I did some benchmarking and cannot get more than 8k qps on my 24 core dev boxes and wondering do you have some tuning instructions. I also did a noop resource which decouples kafka producer side, but  I still cannot get more than 10k qps. Meanwhile I get up a blank http servlet and it could get 55k qps on same machine. 

Best,

Xiang

**Comment by hachikuji** [comment]
Hey @fx19880617 , can you tell us a little more about the tests? Were you just hitting the produce API? What was the structure of the data?

**Comment by fx19880617** [comment]
Hi,
Below is my sample java code to send call to proxy, data is some random load binary(1k bytes).
{  "records": [{"value": "${LOAD_BYTES}"}]}

HttpClient client = new DefaultHttpClient();
HttpPost post = new HttpPost("http://" + hostname + "/topics/" + topicName);
post.setHeader("Content-Type", "application/vnd.kafka.binary.v1+json");
post.setEntity(loadEntity);
HttpResponse response = client.execute(post);

And then I add a noop method into TopicsResource.java, and getting no more than 10k qps.
  @POST
  @Path("/{topic}")
  @Consumes({Versions.KAFKA_V1_JSON_NOOP})
  public void produceNoop(final @Suspended AsyncResponse asyncResponse,
      @PathParam("topic") String topicName,
      @Valid TopicProduceRequest<JsonTopicProduceRecord> request) {
    ProduceRequestCallback callback = new ProducerPool.ProduceRequestCallback() {
      @Override
      public void onCompletion(Integer keySchemaId, Integer valueSchemaId,
          List<RecordMetadataOrException> results) {
        ProduceResponse response = new ProduceResponse();
        response.setOffsets(new Vector<PartitionOffset>());
        response.setKeySchemaId(0);
        response.setValueSchemaId(0);
        asyncResponse.resume(response);
      }
    };
    callback.onCompletion(0, 0, null);
  }

**Comment by hachikuji** [comment]
Thanks! And did the servlet that you used for comparison accept the same payload (i.e. JSON serialized with base64 binary data)?

**Comment by fx19880617** [comment]
Correct, but for that servlet I just read the length of the payload and make sure I do get the load correctly.
This is kind of comparison from the noop,

**Comment by fx19880617** [comment]
Btw, I also test the load with 10k bytes, the performance deceasing is about 5%.

**Comment by hachikuji** [comment]
Thanks @fx19880617 , would you mind posting your servlet handler as well? We'll try to reproduce your test results and see what we can speed up. By the way, were you working off of version 1.0 or master? There was an issue affecting throughput with the 1.0 release (see https://github.com/confluentinc/rest-utils/pull/14), so you might try your test again on master (if you haven't already).

**Comment by fx19880617** [comment]
Hi,

I'm using the latest code.

For the servlet, I'm using this for simple test:

package com.uber.stream.core.kafka.rest.servlet;

import java.io.IOException;
import java.io.PrintWriter;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

public class KafkaTestRestServlet extends HttpServlet {

  protected void doPost(HttpServletRequest req, HttpServletResponse resp)
      throws ServletException, IOException {
    String str = req.getReader().readLine();
    resp.setContentType("text/html");
    PrintWriter out = resp.getWriter();

    out.println(str.length());

    out.close();
  }
}


////////
package com.uber.stream.core.kafka.rest.servlet;

import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder; 
 

public class Main {

  
  
  public static void main(String[] args) throws Exception { 
    
    int port = Integer.parseInt(args[0]);
    
    Server server = new Server(port);
    
    ServletContextHandler context = new ServletContextHandler(ServletContextHandler.SESSIONS);
    context.setContextPath("/");
    server.setHandler(context);

    context.addServlet(new ServletHolder(new KafkaRestServlet()),"/*"); 
    context.addServlet(new ServletHolder(new KafkaRestServlet()),"/topics/*"); 

    server.start();
    server.join();
    
  }
  
  
}

**Comment by ewencp** [comment]
@fx19880617 I think you're just seeing the overhead of all the processing that Jersey does, which is additional stuff on top of the basic servlet HTTP request processing. You can verify this for yourself by trying a small app consisting of just a JAX-RS resource. Here's a small example I tested with that mirrors your servlet example:

```
@Path("/")
public class TestResource {

  @POST
  public String test(String body) {
    return body;
  }

  public static void main(String[] args) throws Exception {
    int port = 8082;
    Server server = new Server(port);

    ServletContextHandler context = new ServletContextHandler(ServletContextHandler.SESSIONS);
    context.setContextPath("/");
    server.setHandler(context);

    ResourceConfig resourceConfig = new ResourceConfig();
    resourceConfig.register(TestResource.class);
    ServletContainer servletContainer = new ServletContainer(resourceConfig);
    context.addServlet(new ServletHolder(servletContainer), "/*");

    server.start();
    server.join();
  }

}
```

Instead of something homegrown, I'd recommend using a well-tested benchmark tool like `ab` as a client. I used a command line similar to this (with adjustments to the URL and data in `postfile` depending on the exact test):

```
ab -t 15 -k -T "application/vnd.kafka.binary.v1+json" -p postfile http://localhost:8082/topics/test
```

JAX-RS does a lot more than what your simple servlet does, including at least:

* content negotiation (Content-Type and Accept headers)
* parsing of request URLs and translation to resource classes' methods, and translation of parts of those URLs to method parameters
* a bunch of additional error handling (i.e. exception handlers)

We also include a bunch of standard stuff in rest-utils (and therefore kafka-rest), some of which depends on JAX-RS hooks, including:

* request logging (a la Apache or Common Log Format logs)
* metrics, including per-request latency, throughput, request/response size, and tracking of TCP-connection level stats
* parsing of the request body from JSON, including validation with Hibernate. For binary data this includes doing the base64 decoding.

There's definitely some overhead to using JAX-RS, but it also drastically simplifies the REST proxy implementation. I think there are probably a lot of performance improvements to be made, but I'm not sure a plain servlet is the best starting point for comparison. For reference, on my Macbook, I get somewhere around 18k requests per second with your plain servlet example, around 6k requests per second with a simple JAX-RS resource, and depending on which features are used, down to ~ 4k requests per second with a nearly barebones implementation of a resource method (and even then, the results are quite noisy, so take those numbers with a grain of salt). So it looks like most of the overhead is going to stuff that Jersey is always going to do for every request.

Perhaps it would be more helpful to discuss the particular type of workload you're working with? Is there a reason you need a very high number of requests per second? Since HTTP requests have fairly high overhead and since the requests are synchronous, usually I would expect REST proxy clients to be a bit more aggressive with batching to reduce the impact of that overhead. Or is it just that we're currently not scaling well to the number of cores you have available, and better parallelism would improve the throughput to levels you're expecting to see across 24 cores?

**Comment by fx19880617** [comment]
Thanks @ewencp for your kindly reply and help to confirm where the overhead comes from.

For our use case, we have thousands app boxes are trying to produce kafka events, it's very high throughput logging (hundreds k qps) plus we also have many non-java applications. So we want to extract the rest-proxy layer out, and want to test the max throughput here.

So that I want to reduce those overhead as much as possible.

I also checked the number for internal LNKD kafka-proxy implementation, they are using http servlet and could scale up to 35-50k qps.

Btw, If I start multiple processes for rest-proxy with different port, it could scale up.

Really want to see if there is more performance tuning for this.

**Comment by ewencp** [comment]
Are you doing any batching before making requests to the REST proxy? Batching will improve throughput and reduce the overhead of the HTTP requests. The producer in the REST proxy will do batching for Kafka, but the same reasoning applies to requests sent to the REST proxy, and probably more so since there's more overhead to an HTTP request compared to a produce request to a broker.

You shouldn't need to start additional processes to scale up. With the setup I tested with, if you just add the `-c` flag to `ab`, you can see the REST proxy start to use more cores. The default Jetty thread pool should scale up to a large number of threads.

Since there's overhead just from using Jersey instead of a plain servlet, that's a good place to look to tweak settings. One relatively easy thing you could do is update the `jersey.version` property in the POM. You might need to do this both in rest-utils and kafka-rest. Currently we're on older versions of Jetty and Jersey because we still support Java 6. If you can use a newer version of Java, there has been some performance work on more recent versions (see the general trend [here](https://blogs.oracle.com/japod/entry/jersey_2_performance), [this](http://blog.dejavu.sk/2015/02/12/performance-improvements-of-sub-resource-locators-in-jersey/) is one example of the type of stuff they are improving, although that particular fix will not help here since we do not use subresource locators). This requires very minor changes due to an interface that changed, but should be easy to test. My quick test showed that the trivial JAX-RS example I gave above went from 6k to 9k requests per second. There may also be other tweaks you could make to the Jersey server configuration (i.e. things that aren't specific to kafka-rest), such as disabling some features, although I haven't dug into which of these might help.

**Comment by fx19880617** [comment]
Not really batching as we have thousands clients will call REST proxy.
I already tried to upgrade jersey or jetty version, but no significant help.

**Comment by fx19880617** [comment]
Also tried to follow this(http://wiki.eclipse.org/Jetty/Howto/High_Load), but still no big help :)

**Comment by ewencp** [comment]
@fx19880617 What jersey and jetty versions? And did you change it in rest-utils? I think the `jersey.version` setting in kafka-rest's pom.xml is just accidentally leftover -- it's not actually referenced anywhere. I'm just wondering since I saw a pretty significant jump moving from Jersey 2.6 to 2.16. How much difference did you actually observe? Given I saw a 50% improvement just by compiling against a newer version + running under Java 7, I'd be surprised if you didn't see *some* increase. Also, what's your exact methodology? There's a bit of a ramp up period before performance actually peeks; before that period the results are noisy for me.

The Jetty wiki page on high load probably won't help much since a lot of it's tweaks focus on the network stack. I think the limit you're currently encountering is due to CPU (which is why parallelizing does help).

For the batching, I was wondering because in terms of requests per second, depending on the rate at which clients are generating events, the difference between no batching and waiting 10ms to accumulate more data can easily be a factor of 10-100 in terms of total requests per second.

Using Jersey certainly isn't a requirement, but does make the code a lot simpler since it deals with a lot of the basics automatically. If there's a compelling enough performance argument, we could work around that, but I'd want to dig into some profiling data before making any decision like that. Right now it's unclear what in Jersey is eating CPU.
